### PR TITLE
Content Helper Sidebar: Add Title Suggestions feature

### PR DESCRIPTION
**Pull request by vaurdan**
## Description
As requested in #1886, implement a feature to generate a number of titles based on the post content. With this solution, currently it's only possible to generate titles for the entire post, but the foundations for allowing to generate titles for blocks of texts - alas headings - are there.

## Motivation and context
Give more power to the editor to have suggested titles based on their website data and on historical data of titles that performed well.

## How has this been tested?
Tested locally on my development instance, and I'm also including integration tests for the Content Suggestions API request.

## Screenshots (if appropriate)
Not a screenshot, but a video:

https://github.com/Parsely/wp-parsely/assets/5804923/21889a53-57fb-4dd4-a07b-87051a123341

### CODE REVIEW

File: phpstan.neon [hunk 1]
```diff
@@ -13,7 +13,7 @@ parameters:
     - vendor/php-stubs/wordpress-stubs/wordpress-stubs.php
     - vendor/php-stubs/wordpress-tests-stubs/wordpress-tests-stubs.php
   type_coverage:
-      return_type: 93
```
Comment (vaurdan) [inline phpstan.neon:16]: PHPStan was reporting that 93.0% of types is below the target of 93% ¯\_(ツ)_/¯

**Comment by acicovic** [comment]
Hey @vaurdan, thanks for working on this, it's a very important feature. I haven't reviewed the code yet but noticed that we've got some [SonarCloud warnings](https://sonarcloud.io/summary/new_code?id=Parsely_wp-parsely&pullRequest=1967).

Can you look into them? If they are false positives or insignificant, we can flag them as `wontfix`.

**Comment by vaurdan** [comment]
@acicovic thanks for the warning! I have reviewed the warnings, and marked the variable warning in the WriteAPITest as wontfix - since it's expected, and doesn't require any fixing - and also marked the usage of a insecure (non HTTPS) API as Safe, given that - for now- the API only responds to HTTP requests. Once we have a secure alternative, I will update the API URL to use it instead.

**Comment by acicovic** [comment]
Hey @vaurdan, I took the liberty to take care of some merge conflicts with the `develop` branch and fix a PHPStan error. The code now passes linting without any issues on my end and it seems to function as expected.

I've also given the UI a manual test. The UX is great and I think the way you've implemented undo is fantastic. I'm wondering whether it would be a good idea to totally remove the `x` button when a title is pinned, and whether this would be easy to do. I'd appreciate if you had any feedback regarding that and if you think it's a good idea or not.

I'll look into the specifics of the code soon!

**Comment by acicovic** [comment]
Very minor UI glitch in the undo functionality: When the title is very short, the border touches the button.

image

File: phpstan.neon (hunk 1 above)
Comment (acicovic) [reply phpstan.neon:16]: Can you try again with `93` with the latest `develop` merge? I tested it and 93% works for me.

**Comment by vaurdan** [comment]
@acicovic thanks for your feedback!

>  I'm wondering whether it would be a good idea to totally remove the x button when a title is pinned, and whether this would be easy to do. I'd appreciate if you had any feedback regarding that and if you think it's a good idea or not.

I thought about that, and should be relatively simple to implement. I ended up going with the disable option as it was the simplest approach, but actually removing/hiding the button might help uncluttering the UI a little bit, so we can try that and see how it looks!

> Very minor UI glitch in the undo functionality: When the title is very short, the border touches the button.

Ah! Well catched. It might need some CSS magic to fix that padding when the container is that small.

**Comment by acicovic** [comment]
Today I've started working on polishing the feature for release, so @vaurdan can focus on other work. 

Some internal changes:
- Use HTTPS for our endpoint
- Remove `ts-ignore` comments

Some UI changes:

- Update the UI's strings
- Use a subtler gray for the pin button
- Minor readjust to title buttons positioning
- Adjustments to the title replacement prompt
- Remove the close button when the title is pinned

I think that the changes made are improving the experience, but if not we can always rollback specific commits. Other than that, there's still some work to do and I think I might need a couple of days to complete it. After that, we should be nearly ready to merge this.
 
image
image
image

**Comment by vaurdan** [comment]
That looks so good @acicovic! Thanks for working on those improvements!

**Review by acicovic — APPROVED** [review]
We're ready to merge this!

We've implemented error handling and some final UI adjustments. This is how the panel looks with the final adjustments:

image

**Pull request merged**